core/src/main/scala/chisel3/Data.scala

```diff
@@ -214,8 +214,28 @@ private[chisel3] object getRecursiveFields {
 private[chisel3] object getMatchedFields {
   def apply(x: Data, y: Data): Seq[(Data, Data)] = (x, y) match {
     case (x: Element, y: Element) =>
-      require(x.typeEquivalent(y))
+      require(
+        x.typeEquivalent(y), {
+          val reason = x
+            .findFirstTypeMismatch(y, strictTypes = true, strictWidths = true, strictProbeInfo = true)
+            .map(s => s"\nbecause: $s")
+            .getOrElse("")
+          s"$x is not typeEquivalent to $y $reason"
+        }
+      )
```

Comment:
We have written essentially the same thing like 5 times. Note that `typeEquivalent` is actually just `findFirstTypeMismatch(...).isEmpty` so we should rewrite this into a utility, something like:
```scala
def requireTypeEquivalent(a: Data, b: Data, strictTypes: Boolean, strictWidths: Boolean, strictProbeInfo: Boolean, extraMsg: => Option[String] = None, err: String => Unit = msg => throwException(msg)) = ...
```